src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/RoundaboutMissingTagCheck.java

```diff
@@ -0,0 +1,196 @@
+    private static final int MODULUS = 10;
+    private static final int FIRST_EDGE_SECTION = 1;
+    private static final List<String> FALLBACK_INSTRUCTIONS = Arrays
+            .asList(MISSING_JUNCTION_TAG_INSTRUCTION);
+    private static final double MAX_THRESHOLD_DEGREES_DEFAULT = 40.0;
+    private static final double MIN_THRESHOLD_DEGREES_DEFAULT = 10.0;
+    private final Angle maxAngleThreshold;
+    private final Angle minAngleThreshold;
+
+    /**
+     * Default constructor
+     *
+     * @param configuration
+     *            the JSON configuration for this check
+     */
+    public RoundaboutMissingTagCheck(final Configuration configuration)
+    {
+        super(configuration);
+        this.maxAngleThreshold = configurationValue(configuration, "threshold.degrees",
+                MAX_THRESHOLD_DEGREES_DEFAULT, Angle::degrees);
+        this.minAngleThreshold = configurationValue(configuration, "threshold.degrees",
+                MIN_THRESHOLD_DEGREES_DEFAULT, Angle::degrees);
+    }
+
+    /**
+     * Validates if given {@link AtlasObject} is actually an {@link Edge} and is a potential
+     * roundabout and also corresponding OSM identifier shouldn't be flagged before (this is for
+     * avoiding duplicate flags)
+     */
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return object instanceof Edge && !this.isFlagged(object.getOsmIdentifier());
+    }
+
+    /**
+     * Flags an {@link Edge} is its circular shape and connected to at least
+     * {@link RoundaboutMissingTagCheck#MINIMUM_INTERSECTION} navigable roads. {@link Edge} doesn't
+     * have junction=roundabout or area=yes tags.
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Edge edge = (Edge) object;
+
+        // process only first sectioned Edge
+        if (edge.getIdentifier() % MODULUS == FIRST_EDGE_SECTION && edge.isMasterEdge()
+                && HighwayTag.isCarNavigableHighway(edge) && edge.getTag(JunctionTag.KEY).isEmpty()
+                && edge.getTag(AreaTag.KEY).isEmpty() && isPartOfClosedWay(edge)
+                && intersectingWithMoreThan(edge))
+        {
+            // rebuild original OSM Way geometry
+            final PolyLine originalGeom = buildOriginalOsmWayGeometry(edge);
+            // check maximum angle
+            final List<Tuple<Angle, Location>> maxOffendingAngles = originalGeom
+                    .anglesGreaterThanOrEqualTo(this.maxAngleThreshold);
+            // check minimum angle
+            final List<Tuple<Angle, Location>> minOffendingAngles = originalGeom
+                    .anglesLessThanOrEqualTo(this.minAngleThreshold);
+
+            if (maxOffendingAngles.isEmpty() && minOffendingAngles.isEmpty())
+            {
+                this.markAsFlagged(object.getOsmIdentifier());
+                return Optional.of(createFlag(new OsmWayWalker(edge).collectEdges(),
```

Comment:
You use the way walker 3 times to collect the entire way (here, in buildOriginalOsmWayGeometry, and in intersectingWithMoreThan). While this is not a super expensive function, its repeated use can add up. It might be a bit better to call it once and then pass the edge set to each of the places that need it.

Reply:
Thanks for suggestion!

Reply:
compilation time. tested against China data:
1 - use WayWalker in each method - 7,48 mi (current)
2 - call WayWalker only ones and pass Set<Edge> edges to every method - 9,64 min
3 - make Set<Edge> static - 8,50 minutes.